sway/tree/workspace.c

```diff
@@ -743,6 +725,49 @@ void workspace_add_gaps(struct sway_workspace *ws) {
 		ws->current_gaps.right += ws->gaps_inner;
 		ws->current_gaps.bottom += ws->gaps_inner;
 		ws->current_gaps.left += ws->gaps_inner;
+	} else {
+		// For tiled containers we need to add the half-gap all around the edge
+		// to match the half gaps that the children have all already added around
+		// themselves. We use the opposite order for the two halves so that the
+		// sum adds up to the correct total gap size in all circumstances.
+		int gaps_horizontal1 = ws->gaps_inner / 2;
+		int gaps_horizontal2 = ws->gaps_inner - gaps_horizontal1;
+		ws->current_gaps.left += gaps_horizontal2;
+		ws->current_gaps.right += gaps_horizontal1;
+
+		int gaps_vertical1 = ws->gaps_inner / 2;
+		int gaps_vertical2 = ws->gaps_inner - gaps_vertical1;
+		ws->current_gaps.top += gaps_vertical2;
+		ws->current_gaps.bottom += gaps_vertical1;
+	}
+
+	// Prevent the resulting gaps from turning out negative, since outer gaps can
+	// be negative and nothing ensures they will not be larger than the current
+	// inner gaps.
+	int mingaps = ws->layout == L_TABBED ||
+	              ws->layout == L_STACKED ? 0 : -ws->gaps_inner;
+	ws->current_gaps.left = MAX(mingaps, ws->current_gaps.left);
```

Comment:
I don't think this is right. The outer gaps should be able to negate the inner gaps for that side, but this would allow for it to negate double the inner gaps. I think these should all be like `ws->current_gaps.left = MAX(0, ws->current_gaps.left)`. 

_It looks like this function was just missing a call to `prevent_invalid_outer_gaps` in between setting the outer gaps and inner gaps._

Reply:
How do you get a double negation? I've gotten rid of `prevent_invalid_outer_gaps` completely, it doesn't seem workable to cover all cases with that. It just does that calculation on apply instead. So it allows the outer gaps to at most be `-ws->gaps_inner` for tiled containers to then allow those to be negated by the child gaps. It also seems to make more sense to not manipulate the values the user actually sets himself as the workspace may change sizes by moving from one output to another.

This reminds me of another corner case. If the gaps are large enough that the gap clamping will apply in the children the negative gap will not be properly negated... That doesn't seem trivial to fix.

Reply:
> How do you get a double negation?

Any values where outer `outer == -2 * inner` should work, but the following should make it easily visible
```
swaymsg -- gaps outer all set -400, gaps inner all set 200
```

> So it allows the outer gaps to at most be `-ws->gaps_inner`

Currently, it is being checked after inner gaps are calculated so it actually makes sure that *all* gaps are at most `-ws->gaps_inner`, which is incorrect.

> This reminds me of another corner case. If the gaps are large enough that the gap clamping will apply in the children the negative gap will not be properly negated

Can you try explaining that further? When would that occur?

Reply:
> Currently, it is being checked after inner gaps are calculated so it actually makes sure that all gaps are at most `-ws->gaps_inner`, which is incorrect.

It's correct because the children will then apply `ws->gaps_inner` themselves and thus negate that.

> Can you try explaining that further? When would that occur?

Set outer gaps to -1000 and inner gaps to +1000.  Now start opening windows. At first it's correct with no outer gaps but gaps within windows. As you open more windows and the inner gaps get clamped the windows start falling outside the screen. This is going to be annoying to fix...

Reply:
It's still wrong though, because clamping needs to be to half size only per edge...

But this doesn't work generally because of the other bug. I'll have to think of a better solution.

Reply:
I've pushed a fix for the half-sized per edge, but this is not mergeable like this because of the other bug. It also doesn't even work always like this because since after the gap clamping it can be negative, the code that adjusts outer gaps can return broken results.